### PR TITLE
add train_step, fit, and compile to the tutorial

### PR DESCRIPTION
**Pull request by haifeng-jin**

### CODE REVIEW

File: guides/intro_to_keras_for_researchers.py [hunk 1]
```diff
@@ -705,11 +705,64 @@ def call(self, inputs, training=None):
...
+    optimizer=keras.optimizers.Adam(learning_rate=1e-3),
+    metrics=[keras.metrics.SparseCategoricalAccuracy()])
+
+# Train the model with the dataset for 2 epochs.
+model.fit(dataset, epochs=2)
+model.evaluate(dataset)
+
+"""
+You can always subclass the `Model` class (it works exactly like subclassing
+`Layer`) if you want to leverage these loops for your OO models.  You can
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:730]: if you want to leverage built-in training infrastructure for your own training loop, such as:

- Callbacks (link)
- Distributed training support (link)
- Step fusing (link)

File: guides/intro_to_keras_for_researchers.py [hunk 2]
```diff
@@ -705,11 +705,64 @@ def call(self, inputs, training=None):
...
+override `train_step` to customize what happens in the training loop. The
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:731]: `train_step()`

File: guides/intro_to_keras_for_researchers.py [hunk 3]
```diff
@@ -705,11 +705,64 @@ def call(self, inputs, training=None):
...
+model = tf.keras.Model(inputs, outputs)
+
+# Specifying the loss, optimizer, and metrics with `compile()`.
+model.compile(
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    optimizer=keras.optimizers.Adam(learning_rate=1e-3),
+    metrics=[keras.metrics.SparseCategoricalAccuracy()])
+
+# Train the model with the dataset for 2 epochs.
+model.fit(dataset, epochs=2)
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:725]: Let's answer: why is it a good idea to use the built-in fit?

File: guides/intro_to_keras_for_researchers.py (hunk 1 above)
Comment (fchollet) [reply guides/intro_to_keras_for_researchers.py:730]: Let's define these features in the section above and mention them again here

File: guides/intro_to_keras_for_researchers.py [hunk 4]
```diff
@@ -705,11 +705,64 @@ def call(self, inputs, training=None):
...
+        # on what you pass to `fit()`.
+        x, y = data
+        with tf.GradientTape() as tape:
+            y_pred = self(x, training=True)  # Forward pass
+            loss = self.compiled_loss(y, y_pred,
+                                      regularization_losses=self.losses)
+        gradients = tape.gradient(loss, self.trainable_weights)
+        self.optimizer.apply_gradients(zip(gradients, self.trainable_weights))
+        # Update metrics (includes the metric that tracks the loss)
+        self.compiled_metrics.update_state(y, y_pred)
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:749]: compiled_metrics -> never referenced. Don't use

File: guides/intro_to_keras_for_researchers.py [hunk 5]
```diff
@@ -705,11 +705,64 @@ def call(self, inputs, training=None):
...
+        # Return a dict mapping metric names to current value
+        return {m.name: m.result() for m in self.metrics}
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:751]: self.metrics -> same

File: guides/intro_to_keras_for_researchers.py [hunk 6]
```diff
@@ -705,11 +705,64 @@ def call(self, inputs, training=None):
...
+You can always subclass the `Model` class (it works exactly like subclassing
+`Layer`) if you want to leverage these loops for your OO models.  You can
+override `train_step` to customize what happens in the training loop. The
+`train_step` function is called with every batch of training data. The `fit`
+function will compile the training code with `tf.function` by default to make
+it faster.
+"""
+
+class CustomModel(keras.Model):
+    def train_step(self, data):
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:738]: Add an `__init__()` and a `compile(optimizer)`

in `__init__()` define a `metrics.Mean(name="loss")`

File: guides/intro_to_keras_for_researchers.py [hunk 7]
```diff
@@ -705,11 +705,64 @@ def call(self, inputs, training=None):
...
+        # Unpack the data. Its structure depends on your model and
+        # on what you pass to `fit()`.
+        x, y = data
+        with tf.GradientTape() as tape:
+            y_pred = self(x, training=True)  # Forward pass
+            loss = self.compiled_loss(y, y_pred,
+                                      regularization_losses=self.losses)
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:745]: Add comment about self.losses

File: guides/intro_to_keras_for_researchers.py [hunk 8]
```diff
@@ -705,11 +705,64 @@ def call(self, inputs, training=None):
...
+
+# Specifying the loss, optimizer, and metrics with `compile()`.
+model.compile(
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    optimizer=keras.optimizers.Adam(learning_rate=1e-3),
+    metrics=[keras.metrics.SparseCategoricalAccuracy()])
+
+# Train the model with the dataset for 2 epochs.
+model.fit(dataset, epochs=2)
+model.evaluate(dataset)
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:726]: Maybe mention predict()

File: guides/intro_to_keras_for_researchers.py [hunk 9]
```diff
@@ -705,11 +705,64 @@ def call(self, inputs, training=None):
...
+x = keras.layers.Dense(32, activation="relu")(x)
+outputs = keras.layers.Dense(10)(x)
+model = CustomModel(inputs, outputs)
+
+model.compile(
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+    optimizer=keras.optimizers.Adam(learning_rate=1e-3),
+    metrics=[keras.metrics.SparseCategoricalAccuracy()])
+
+model.fit(dataset, epochs=2)
```
Comment (fchollet) [inline guides/intro_to_keras_for_researchers.py:764]: Also mention that we can override `test_step()` for a custom `evaluate()` loop or `predict_step()` for a custom `predict()` step.

Comment (fchollet) [reply guides/intro_to_keras_for_researchers.py:764]: (maybe link to the relevant guide)

File: guides/intro_to_keras_for_researchers.py (hunk 7 above)
Comment (haifeng-jin) [reply guides/intro_to_keras_for_researchers.py]: removed `self.losses` together with `compiled_loss`.
Use a `loss_fn` instead.

File: guides/intro_to_keras_for_researchers.py (hunk 6 above)
Comment (haifeng-jin) [reply guides/intro_to_keras_for_researchers.py:772]: Not sure how to add `compile(optimizer)`.
Assign `self.optimizer=...` directly in `__init__()`.
and call `model.compile()` without any arg.

**Review by fchollet — APPROVED** [review]
LGTM, thank you! Added some copyedits.
